### PR TITLE
add section Upgradeable Contracts

### LINKED ISSUE #82: Add section "Upgradeable Contracts"
We should add a section "Upgradeable Contracts"  under "Basics".

ink! currently has those three approaches for upgradeable contracts: https://github.com/paritytech/ink/tree/master/examples/upgradeable-contracts.

We should have a page in our documentation on each, explaining the use cases and how to use them. The result of https://github.com/paritytech/ink/issues/1247 should be incorporated here as well.

If you have ideas for illustrations to accompany these pages: feel free to approach our illustrator to create some in the ink! style.

### PR DESCRIPTION
**Pull request by SkymanOne**
closes #82

### CODE REVIEW

File: docs/basics/upgradeability.md [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+**You should not change the order in which the contract state variables are declared, nor their type!**
+
+Violating the restriction will not prevent a successful compilation,
+but will result in **the mix-up of values** or **failure to read the storage correctly**.
+This can be a result of severe errors in the application utilizing the contract.
+
+If the storage of your contract looks like this:
+```rust
+use ink_lang as ink;
```
Comment (HCastano) [inline docs/basics/upgradeability.md:139]:
```suggestion
```

File: docs/basics/upgradeability.md [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+#[ink(storage)]
+pub struct YourContract {
+    x: u32,
+    y: bool,
+}
+```
+The procedures listed below will make it ***invalid***:
```
Comment (HCastano) [inline docs/basics/upgradeability.md:146]: Not sure how this looks like rendered, but I'd still had a bit of space around
the codeblocks and commentary to make it easier to read in the source

**Review by HCastano — COMMENTED** [review]
Will be a good addition to the docs, thanks!

Can you wrap lines to 90ish characters? Makes it easier to
read the source files

File: docs/basics/upgradeability.md [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+```
+
+Using this pattern, you can introduce other message to your proxy contract.
+Any messages that are not matched on the proxy contract will be forwarded the specified contract address.
+
+# Updating contract code directly
+
+Following [Substrate's runtime upgradeability](https://docs.substrate.io/tutorials/get-started/forkless-upgrade/) philosophy, ink! also supports an easy way to update your contract code via the special function `set_code_hash()`
+
+## Properties
```
[the remark this answers is not in the document]
Comment (SkymanOne) [reply docs/basics/upgradeability.md:102]: I will contact a designer for this

File: docs/basics/upgradeability.md [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+User ---- tx ---> Proxy ----------> Implementation_v0
+                     |
+                      ------------> Implementation_v1
+                     |
+                      ------------> Implementation_v2
+```
+
+
+Our proxy contract will have these 2 storage fields:
+```rust
```
Comment (cmichi) [inline docs/basics/upgradeability.md:40]:
```suggestion

```rust
```

File: docs/basics/upgradeability.md [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+        unreachable!(
+            "the forwarded call will never return since `tail_call` was set"
+        );
+    }
+}
+```
+
+:::tip
+
+Take a look at the selector pattern in the attribute macro: by declaring `selector = _`,
```
Comment (cmichi) [inline docs/basics/upgradeability.md:111]:
```suggestion
Take a look at the selector pattern in the attribute macro: by declaring `selector = _`
```

File: docs/basics/upgradeability.md [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+we specify that all other messages should be handled by this message selector.
+
+:::
+
+Using this pattern, you can introduce other message to your proxy contract.
+Any messages that are not matched on the proxy contract will be forwarded the specified contract address.
```
Comment (cmichi) [inline docs/basics/upgradeability.md:117]:
```suggestion
Any messages that are not matched in the proxy contract will be forwarded to the specified contract address.
```

File: docs/basics/upgradeability.md [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+
+## Replacing Contract Code with `set_code_hash()`
+
+Following [Substrate's runtime upgradeability](https://docs.substrate.io/tutorials/get-started/forkless-upgrade/) 
+philosophy, ink! also supports an easy way to update your contract code via the special function [`set_code_hash()`](https://paritytech.github.io/ink/ink_env/fn.set_code_hash.html)
```
Comment (cmichi) [inline docs/basics/upgradeability.md:122]:
```suggestion
philosophy, ink! also supports an easy way to update your contract code via the special function [`set_code_hash()`](https://paritytech.github.io/ink/ink_env/fn.set_code_hash.html).
```

File: docs/basics/upgradeability.md [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+
+### Properties
+
+- Updates the contract code using `set_code_hash()`. This effectively replaces the code which is executed for the contract address.
+- The other contract needs to be deployed on-chain.
+- State is stored in the storage of the originally instantiated contract.
+
+Just add the following function to the contract your want to upgrade in future.
```
Comment (cmichi) [inline docs/basics/upgradeability.md:130]:
```suggestion
Just add the following function to the contract you want to upgrade in the future.
```

File: docs/basics/upgradeability.md [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+            code_hash, err
+        )
+    });
+    ink_env::debug_println!("Switched code hash to {:?}.", code_hash);
+}
+```
+
+### Storage Compatibility
+
+It is developer's responsibility to ensure that the new contract's storage is compatible with the old version.
```
Comment (cmichi) [inline docs/basics/upgradeability.md:151]:
```suggestion
It is the developer's responsibility to ensure that the new contract's storage is compatible with the storage of the contract that is replaced.
```

File: docs/basics/upgradeability.md [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+
+You should not change the order in which the contract state variables are declared, nor their type!
+
+Violating the restriction will not prevent a successful compilation,
+but will result in **the mix-up of values** or **failure to read the storage correctly**.
+This can be a result of severe errors in the application utilizing the contract.
+
+:::
+
+If the storage of your contract looks like this:
```
Comment (cmichi) [inline docs/basics/upgradeability.md:163]:
```suggestion

If the storage of your contract looks like this:
```

File: docs/basics/upgradeability.md [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+
+```rust
+#[ink(storage)]
+pub struct YourContract {
+    y: bool,
+    x: u32,
+}
+```
+
+Removing existing variable:
```
Comment (cmichi) [inline docs/basics/upgradeability.md:184]:
```suggestion
Removing an existing variable:
```

File: docs/basics/upgradeability.md [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+
+```rust
+#[ink(storage)]
+pub struct YourContract {
+    x: u32,
+}
+```
+
+Changing type of a variable:
```
Comment (cmichi) [inline docs/basics/upgradeability.md:193]:
```suggestion
Changing the type of a variable:
```

**Review by cmichi — COMMENTED** [review]
Thank you! Got some nits.

**Review by cmichi — APPROVED** [review]

**Pull request merged**